core/include/edm/container.hpp

```diff
@@ -51,9 +84,163 @@ namespace traccc {
 	/// The item vector type
         using item_vector = jagged_vector_type< item_t >;
 	
+        /**
+         * @brief The type name of the element view which is returned by various
+         * methods in this class.
+         */
+        using element_view = container_element<
+            header_t,
+            typename item_vector::value_type
+        >;
+
+        /**
+         * @brief The size type of this container, which is the type by which
+         * its elements are indexed.
+         */
+        using size_type = typename header_vector::size_type;
         /// @}
 
+        /**
+         * We need to assert that the header vector and the outer layer of the
+         * jagged vector have the same size type, so they can be indexed using
+         * the same type.
+         */
+        static_assert(
+            std::is_same_v<
+                typename header_vector::size_type,
+                typename item_vector::size_type
+            >,
+            "Size type for container header and item vectors must be the same."
+        );
+
+        /**
+         * @brief Standard two-argument constructor.
+         *
+         * To enforce the invariant that both vectors must be the same size, we
+         * check this in the constructor. This is also checked in release
+         * builds.
+         */
+        container(
+            header_vector && hv,
+            item_vector && iv
+        )
+            : headers(hv)
+            , items(iv)
+        {
+            if (headers.size() != items.size()) {
+                throw std::logic_error("Header and item length not equal.");
+            }
```

Comment:
I'm not sure how such a constructor will fit into making this type usable in both device and host code at the same time...

If a constructor like this does work in device code, we would still need to use an

```c++
assert( headers.size() == items.size() );
```

statement in it, as exceptions are definitely a no go.

Reply:
Hmm, maybe I have misunderstood Beomki's intention here, but these containers are not designed to be used in device code at all, are they? :confused:

Reply:
Have a look at:

https://github.com/acts-project/traccc/blob/main/core/include/edm/container.hpp#L68-L71

The idea, which I had with my code for the cell container, and which Beomki took for this generalisation as well, is that we just specialise this template class with either the host or device vector types of VecMem, and that's how we get the types used on the host and on the device. So all code present in this class would either have to be usable on a device, or be explicitly marked as only being usable on the host.

The latter is not necessarily a bad thing to do. For instance we can definitely have some constructors be host-only. That's fine. But the "regular" functions should be usable in device code as well.

Reply:
Ah, I guess I misunderstood the design here then... Although this really just seems like important a lot of the annoying constraints of the device programming world into the host world. But sure, we can mark the host-only methods as such. In that case, I would say the bounds-checked accessor methods should be host only, as should of course the `reserve` and `push_back` methods (unless we want to use the device growing methods for this). Alternatively it might be worth defining a `device_container` class instead, similar to the design of vecmem.

Reply:
It's going to definitely useful if we can use them on the device side. I am afraid defining host-only and device-only methods would increase code complexity. Most of all I don't see the use cases for this constructor.